Orange/widgets/visualize/owscatterplot.py

```diff
@@ -380,10 +380,16 @@ def update_graph(self, attributes=None, reset_view=True, **_):
         self.graph.zoomStack = []
         if attributes and len(attributes) == 2:
             self.attr_x, self.attr_y = attributes
+        X = BasicStats(self.data[:, self.attr_x], self.attr_x)
+        Y = BasicStats(self.data[:, self.attr_y], self.attr_y)
```

Comment:
I was not suggesting to move this from `update_graph`, just that the BasicStats can be computed once (e.g. in `set_data`) and the results stored for subsequent use.
But now that I think about it - maybe the warnings should be moved somewhere else as well - they depend only on the selection of attributes right? So warnings should be raised when attribute selection changes. Not when somebody selects/deselects "Show class density".

The complexity is of course O(m) for a single call, the question is how many calls are made.